File: backend/llm.py
"""OpenAI-uyumlu bir endpoint'e sozluk sorgusu atan katman.

Gemini, Groq, OpenRouter ve Ollama'nin hepsi ayni /chat/completions sozlesmesini
konustugu icin tek kod yolu yeterli; saglayici degistirmek .env'de uc satir.
"""

import json
import re

import httpx

from . import config
from .schema import LookupResult

SYSTEM_PROMPT = """Sen deneyimli bir sozlukbilimci ve dil ogretmenisin.
Kullanici sana {lang_name} ({lang_native}) dilinde tek bir kelime verecek.
Turkce konusan bir ogrenci icin o kelimenin tam kaydini hazirlayacaksin.

Kurallar:
- SADECE gecerli JSON dondur. Markdown kod bloklari, aciklama, on soz yazma.
- "turkish_meanings", "turkish_explanation" ve butun "note" alanlari TURKCE olacak.
- "native_definition" alani {lang_native} dilinde, o dilin tek dilli sozlugundeki
  gibi yazilacak. Turkce ceviri degil, hedef dilde gercek bir tanim olacak.
- "synonyms" icindeki her kelimenin "note" alaninda, o es anlamlinin asil
  kelimeden hangi nuansla ayrildigini Turkce olarak tek cumleyle acikla.
  Ayni anlama gelen iki kelimeyi farksizmis gibi gosterme.
- Kullanici cekimli/turemis bir bicim yazarsa ("verstanden", "Haeuser"), "lemma"
  alanina sozluk bicimini koy ve "correction_note" ile Turkce olarak durumu belirt.
- Yazim hatasi varsa en olasi dogru kelimeyi "lemma" yap ve yine not dus.
- Kelime o dilde gercekten yoksa {{"found": false, "word": "...", "correction_note":
  "Turkce aciklama"}} dondur.
- "morphology" alanini dile gore doldur: isimlerde artikel ve cogul, fiillerde
  temel cekimler, sifatlarda karsilastirma bicimleri. Ilgisiz alanlari bos birak.
- "cefr" alanina A1/A2/B1/B2/C1/C2 seviyelerinden birini yaz.
- "collocations" alanina kelimenin gercekten sik gectigi kalip ve birliktelikleri
  yaz (ornegin fiil + edat). Ezber icin en degerli kisim burasidir.
- 2-3 ornek cumle ver, her birinin Turkce cevirisiyle.

Doner JSON semasi:
{{
  "word": "kullanicinin yazdigi kelime",
  "found": true,
  "lemma": "sozluk bicimi",
  "correction_note": "",
  "ipa": "/.../",
  "part_of_speech": "hedef dilde tur + parantez icinde Turkcesi",
  "cefr": "B1",
  "morphology": {{"article": "", "plural": "", "verb_forms": "", "other": ""}},
  "turkish_meanings": ["anlam 1", "anlam 2"],
  "turkish_explanation": "kelimenin nerede, nasil kullanildigina dair Turkce not",
  "native_definition": "hedef dilde tanim",
  "synonyms": [{{"word": "...", "note": "Turkce nuans aciklamasi"}}],
  "antonyms": [{{"word": "...", "note": "Turkce not"}}],
  "collocations": ["kalip 1", "kalip 2"],
  "examples": [{{"sentence": "hedef dilde cumle", "translation": "Turkce ceviri"}}]
}}"""


class LLMError(RuntimeError):
    """Saglayiciya ulasilamadi ya da cevap ayristirilamadi."""


def _normalize(data: dict) -> dict:
    """Modelin serbestce urettigi JSON'u semanin bekledigi bicime cekiyor.

    Iki yaygin sapmayi tolere ediyoruz: null gonderilen alanlar (pydantic'te
    varsayilan degil hata olurdu) ve nesne yerine duz metin listesi olarak gelen
    es/zit anlamlilar.
    """
    data = {k: v for k, v in data.items() if v is not None}

    for key in ("synonyms", "antonyms"):
        items = data.get(key)
        if isinstance(items, list):
            data[key] = [
                {"word": item, "note": ""} if isinstance(item, str) else item
                for item in items
                if item
            ]

    examples = data.get("examples")
    if isinstance(examples, list):
        data["examples"] = [
            {"sentence": item, "translation": ""} if isinstance(item, str) else item
            for item in examples
            if item
        ]

    morphology = data.get("morphology")
    if isinstance(morphology, dict):
        data["morphology"] = {k: v for k, v in morphology.items() if v is not None}
    elif morphology is not None and not isinstance(morphology, dict):
        data.pop("morphology", None)

    meanings = data.get("turkish_meanings")
    if isinstance(meanings, str):
        data["turkish_meanings"] = [meanings]

    return data


def _extract_json(text: str) -> dict:
    """Modelin cevabindan JSON govdesini cikarir.

    Bazi modeller istenmese de ```json citleri ekliyor ya da tanimin onune bir
    cumle yaziyor; ikisini de tolere ediyoruz.
    """
    text = text.strip()
    fenced = re.search(r"```(?:json)?\s*(.+?)\s*```", text, re.DOTALL)
    if fenced:
        text = fenced.group(1).strip()

    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    start, end = text.find("{"), text.rfind("}")
    if start != -1 and end > start:
        try:
            return json.loads(text[start : end + 1])
        except json.JSONDecodeError as exc:
            raise LLMError(f"Model gecerli JSON dondurmedi: {exc}") from exc

    raise LLMError("Model cevabinda JSON bulunamadi.")


async def lookup(word: str, language: str) -> LookupResult:
    if not config.LLM_API_KEY:
        raise LLMError(
            "LLM_API_KEY tanimli degil. .env.example dosyasini .env olarak "
            "kopyalayip anahtarini yaz."
        )

    lang = config.LANGUAGES[language]
    system = SYSTEM_PROMPT.format(lang_name=lang["name"], lang_native=lang["native"])

    payload = {
        "model": config.LLM_MODEL,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": word},
        ],
        "temperature": 0.2,
        "response_format": {"type": "json_object"},
    }

    headers = {
        "Authorization": f"Bearer {config.LLM_API_KEY}",
        "Content-Type": "application/json",
    }

    url = f"{config.LLM_BASE_URL}/chat/completions"

    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(url, json=payload, headers=headers)
    except httpx.HTTPError as exc:
        raise LLMError(f"Saglayiciya ulasilamadi: {exc}") from exc

    if response.status_code == 429:
        raise LLMError(
            "Gunluk ucretsiz kota doldu (429). Yarin tekrar dene ya da .env'den "
            "baska bir saglayiciya gec."
        )
    if response.status_code >= 400:
        raise LLMError(f"Saglayici {response.status_code} dondu: {response.text[:300]}")

    try:
        content = response.json()["choices"][0]["message"]["content"]
    except (KeyError, IndexError, ValueError) as exc:
        raise LLMError(f"Beklenmeyen cevap bicimi: {exc}") from exc

    data = _normalize(_extract_json(content))
    data["word"] = data.get("word") or word
    data["language"] = language

    # Modelin uydurdugu fazladan alanlar sessizce dusuyor, eksikler varsayilana
    # donuyor; arayuz her zaman ayni sekli goruyor.
    return LookupResult.model_validate(data)

File: lib/llm.ts
import {
  EMPTY_MORPHOLOGY,
  LANGUAGES,
  type LanguageCode,
  type LookupResult,
  type Term,
} from "./dictionary";

/**
 * Prompt ve cevap ayristirma. Bu dosya saf: ne ag cagrisi yapar ne de ortam
 * degiskeni okur, boylece hem tarayicida hem sunucuda kullanilabilir.
 */

export function systemPrompt(language: LanguageCode): string {
  const { name, native } = LANGUAGES[language];
  return `Sen deneyimli bir sözlükbilimci ve dil öğretmenisin.
Kullanıcı sana ${name} (${native}) dilinde tek bir kelime verecek.
Türkçe konuşan bir öğrenci için o kelimenin tam kaydını hazırlayacaksın.

Kurallar:
- SADECE geçerli JSON döndür. Markdown kod blokları, açıklama, ön söz yazma.
- "turkishMeanings", "turkishExplanation" ve bütün "note" alanları TÜRKÇE olacak.
- "nativeDefinition" alanı ${native} dilinde, o dilin tek dilli sözlüğündeki gibi
  yazılacak. Türkçe çeviri değil, hedef dilde gerçek bir tanım olacak.
- "synonyms" içindeki her kelimenin "note" alanında, o eş anlamlının asıl
  kelimeden hangi nüansla ayrıldığını Türkçe olarak tek cümleyle açıkla.
  Aynı anlama gelen iki kelimeyi farksızmış gibi gösterme.
- Kullanıcı çekimli/türemiş bir biçim yazarsa, "lemma" alanına sözlük biçimini
  koy ve "correctionNote" ile Türkçe olarak durumu belirt.
- Yazım hatası varsa en olası doğru kelimeyi "lemma" yap ve yine not düş.
- Kelime o dilde gerçekten yoksa {"found": false, "word": "...", "correctionNote":
  "Türkçe açıklama"} döndür.
- "morphology" alanını dile göre doldur: isimlerde artikel ve çoğul, fiillerde
  temel çekimler, sıfatlarda karşılaştırma biçimleri. İlgisiz alanları boş bırak.
- "cefr" alanına A1/A2/B1/B2/C1/C2 seviyelerinden birini yaz.
- "collocations" alanına kelimenin gerçekten sık geçtiği kalıpları yaz.
- 2-3 örnek cümle ver, her birinin Türkçe çevirisiyle.

Dönen JSON şeması:
{
  "word": "kullanıcının yazdığı kelime",
  "found": true,
  "lemma": "sözlük biçimi",
  "correctionNote": "",
  "ipa": "/.../",
  "partOfSpeech": "hedef dilde tür + parantez içinde Türkçesi",
  "cefr": "B1",
  "morphology": {"article": "", "plural": "", "verbForms": "", "other": ""},
  "turkishMeanings": ["anlam 1", "anlam 2"],
  "turkishExplanation": "kelimenin nerede, nasıl kullanıldığına dair Türkçe not",
  "nativeDefinition": "hedef dilde tanım",
  "synonyms": [{"word": "...", "note": "Türkçe nüans açıklaması"}],
  "antonyms": [{"word": "...", "note": "Türkçe not"}],
  "collocations": ["kalıp 1", "kalıp 2"],
  "examples": [{"sentence": "hedef dilde cümle", "translation": "Türkçe çeviri"}]
}`;
}

export class LookupError extends Error {}

/**
 * Modelin cevabindan JSON govdesini cikarir. Bazi modeller istenmese de ```json
 * citleri ekliyor ya da tanimin onune bir cumle yaziyor; ikisini de tolere ediyoruz.
 */
export function extractJson(text: string): unknown {
  const trimmed = text.trim();
  const fenced = /```(?:json)?\s*([\s\S]+?)\s*```/.exec(trimmed);
  const candidate = fenced ? fenced[1].trim() : trimmed;

  try {
    return JSON.parse(candidate);
  } catch {
    // devam: gomulu nesneyi aramaya calis
  }

  const start = candidate.indexOf("{");
  const end = candidate.lastIndexOf("}");
  if (start !== -1 && end > start) {
    try {
      return JSON.parse(candidate.slice(start, end + 1));
    } catch (error) {
      throw new LookupError(`Model geçerli JSON döndürmedi: ${String(error)}`);
    }
  }

  throw new LookupError("Model cevabında JSON bulunamadı.");
}

const asString = (value: unknown): string =>
  typeof value === "string" ? value : "";

const asStringArray = (value: unknown): string[] => {
  if (typeof value === "string") return value ? [value] : [];
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string");
};

/** Model bazen nesne yerine duz metin listesi donduruyor; ikisini de kabul et. */
const asTerms = (value: unknown): Term[] => {
  if (!Array.isArray(value)) return [];
  return value.flatMap((item): Term[] => {
    if (typeof item === "string") return item ? [{ word: item, note: "" }] : [];
    if (item && typeof item === "object") {
      const record = item as Record<string, unknown>;
      const word = asString(record.word);
      return word ? [{ word, note: asString(record.note) }] : [];
    }
    return [];
  });
};

export function normalizeResult(
  raw: unknown,
  word: string,
  language: LanguageCode,
): LookupResult {
  const data = (raw ?? {}) as Record<string, unknown>;
  const morphologyRaw = (
    data.morphology && typeof data.morphology === "object"
      ? data.morphology
      : {}
  ) as Record<string, unknown>;

  const examplesRaw = Array.isArray(data.examples) ? data.examples : [];

  return {
    word: asString(data.word) || word,
    language,
    found: data.found !== false,
    lemma: asString(data.lemma),
    correctionNote: asString(data.correctionNote),
    ipa: asString(data.ipa),
    partOfSpeech: asString(data.partOfSpeech),
    cefr: asString(data.cefr),
    morphology: {
      ...EMPTY_MORPHOLOGY,
      article: asString(morphologyRaw.article),
      plural: asString(morphologyRaw.plural),
      verbForms: asString(morphologyRaw.verbForms),
      other: asString(morphologyRaw.other),
    },
    turkishMeanings: asStringArray(data.turkishMeanings),
    turkishExplanation: asString(data.turkishExplanation),
    nativeDefinition: asString(data.nativeDefinition),
    synonyms: asTerms(data.synonyms),
    antonyms: asTerms(data.antonyms),
    collocations: asStringArray(data.collocations),
    examples: examplesRaw.flatMap((item) => {
      if (typeof item === "string") return [{ sentence: item, translation: "" }];
      if (item && typeof item === "object") {
        const record = item as Record<string, unknown>;
        const sentence = asString(record.sentence);
        return sentence
          ? [{ sentence, translation: asString(record.translation) }]
          : [];
      }
      return [];
    }),
  };
}
